tests/robustness/validate/watch.go

```diff
@@ -138,8 +141,44 @@ func validateResumable(t *testing.T, events []model.WatchEvent, report report.Cl
 		}
 		firstEvent := firstWatchEvent(op)
 		// If watch is resumable, first event it gets should the first event that happened after the requested revision.
-		if firstEvent != nil && events[index] != *firstEvent {
-			t.Errorf("Resumable - A broken watch can be resumed by establishing a new watch starting after the last revision received in a watch event before the break, so long as the revision is in the history window, watch request: %+v, event missing: %+v", op.Request, events[index])
+		if firstEvent != nil && !events[index].Equals(*firstEvent) {
+			t.Errorf("Resumable - A broken watch can be resumed by establishing a new watch starting after the last revision received in a watch event before the break, so long as the revision is in the history window, watch request: %+v, event missing: %+v, got: %+v", op.Request, events[index], *firstEvent)
+		}
+	}
+}
+
+// validatePrevKV ensures that a watch response (if configured with WithPrevKV()) returns
+// the appropriate response.
+func validatePrevKV(t *testing.T, report report.ClientReport, history []model.WatchEvent) {
+	replay := model.NewReplay(history)
+	for _, op := range report.Watch {
+		for _, resp := range op.Responses {
+			for _, event := range resp.Events {
+				// get state state just before the current event.
+				state, err := replay.StateForRevision(event.Revision - 1)
+				if err != nil {
+					t.Error(err)
+				}
+
+				// It is not nescessary that PrevValue will always be nil for a current event.
+				// However, if PrevValue is nil provided a watch is opened with PrevKV configured,
+				// then that has to be a create event.
+				//
+				// Furthermore, this explicitly checks the condition that is used by Kubernetes
+				// while parsing events received from etcd:
+				// https://github.com/kubernetes/kubernetes/blob/a9e4f5b7862e84c4152eabe2e960f3f6fb9a4867/staging/src/k8s.io/apiserver/pkg/storage/etcd3/event.go#L59
+				if event.PrevValue == nil && !event.CreateEvent {
+					t.Errorf("PrevKV - got a create event with nil prevKV %+v", event)
+				}
```

Comment:
@serathius thinking about this a little more, could use your input:

I'm not sure anymore if we should be checking this condition anymore here. Looks like in the face of compaction, prevKV can also be nil: 
https://github.com/etcd-io/etcd/blob/43d675997772bee9cf9d6bd4804ef05cc0cecb01/api/etcdserverpb/rpc.proto#L799-L801

Reply:
It's good point, please skip it for now as we don't compact in robustness tests for now.  

I'm not sure documentation is correct as Kubernetes contract correctness depends on this field being available. Will investigate it further.

Reply:
The discussion here is most relevant: https://github.com/kubernetes/kubernetes/issues/76624

Reply:
Right.